urllib3/util/url.py

```diff
@@ -1,37 +1,80 @@
 from __future__ import absolute_import
-from collections import namedtuple
 
 from ..exceptions import LocationParseError
+from ..packages import rfc3986
+from ..packages import six
 
 
 url_attrs = ['scheme', 'auth', 'host', 'port', 'path', 'query', 'fragment']
 
 
-class Url(namedtuple('Url', url_attrs)):
+class Url(object):
     """
     Datastructure for representing an HTTP URL. Used as a return value for
     :func:`parse_url`. Both the scheme and host are normalized as they are
     both case-insensitive according to RFC 3986.
     """
-    __slots__ = ()
 
-    def __new__(cls, scheme=None, auth=None, host=None, port=None, path=None,
-                query=None, fragment=None):
+    def __init__(self, scheme=None, auth=None, host=None, port=None, path=None,
+                 query=None, fragment=None, parse_result=None):
         if path and not path.startswith('/'):
             path = '/' + path
-        if scheme:
-            scheme = scheme.lower()
-        if host:
-            host = host.lower()
-        return super(Url, cls).__new__(cls, scheme, auth, host, port, path,
-                                       query, fragment)
+            if parse_result is not None:
+                parse_result = parse_result.copy_with(path=path)
```

Comment:
Do we want to move this logic into a custom constructor, rather than overload the default constructor?